ParOpt is failing to build for all test configurations

Building with ParOpt (option `-a`) is failing on Ubuntu and MacOS baseline configurations as well as Ubuntu Latest:
```
----------------------- Building PAROPT from source code ----------------------
Cloning https://github.com/smdogroup/paropt.git...
 
Changed directory to /tmp/tmp_nnp492g
 
 
Installing Cython with pip...
Requirement already satisfied: Cython in /usr/share/miniconda/envs/test/lib/python3.10/site-packages (3.0.3)
 
Building...

gmake[1]: Entering directory '/tmp/tmp_nnp492g/src'
mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c ParOptOptions.cpp -o ParOptOptions.o

        --- Compiled ParOptOptions.cpp successfully ---

mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c ParOptInteriorPoint.cpp -o ParOptInteriorPoint.o

        --- Compiled ParOptInteriorPoint.cpp successfully ---

mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c ParOptVec.cpp -o ParOptVec.o

Traceback (most recent call last):
  File "/usr/share/miniconda/envs/test/bin/build_pyoptsparse", line 8, in <module>
    sys.exit(perform_install())
  File "/usr/share/miniconda/envs/test/lib/python3.10/site-packages/build_pyoptsparse.py", line 1314, in perform_install
    install_pyoptsparse_from_src()
  File "/usr/share/miniconda/envs/test/lib/python3.10/site-packages/build_pyoptsparse.py", line 900, in install_pyoptsparse_from_src
    install_paropt_from_src()
  File "/usr/share/miniconda/envs/test/lib/python3.10/site-packages/build_pyoptsparse.py", line [707](https://github.com/swryan/build_pyoptsparse/actions/runs/6482776210/job/17602952498#step:9:708), in install_paropt_from_src
    pip_install(['./'], pkg_desc='paropt')
  File "/usr/share/miniconda/envs/test/lib/python3.10/site-packages/build_pyoptsparse.py", line 491, in pip_install
    run_cmd(cmd_list)
  File "/usr/share/miniconda/envs/test/lib/python3.10/site-packages/build_pyoptsparse.py", line 402, in run_cmd
    raise inst
  File "/usr/share/miniconda/envs/test/lib/python3.10/site-packages/build_pyoptsparse.py", line 399, in run_cmd
    result = subprocess.run(cmd_list, check=do_check, capture_output=True, text=True)
  File "/usr/share/miniconda/envs/test/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['python', '-m', 'pip', 'install', './']' returned non-zero exit status 1.
        --- Compiled ParOptVec.cpp successfully ---

mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c ParOptQuasiNewton.cpp -o ParOptQuasiNewton.o

        --- Compiled ParOptQuasiNewton.cpp successfully ---

mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c ParOptMMA.cpp -o ParOptMMA.o

        --- Compiled ParOptMMA.cpp successfully ---

mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c ParOptTrustRegion.cpp -o ParOptTrustRegion.o

        --- Compiled ParOptTrustRegion.cpp successfully ---

mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c ParOptProblem.cpp -o ParOptProblem.o

        --- Compiled ParOptProblem.cpp successfully ---

mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c ParOptOptimizer.cpp -o ParOptOptimizer.o

        --- Compiled ParOptOptimizer.cpp successfully ---

mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c ParOptCompactEigenvalueApprox.cpp -o ParOptCompactEigenvalueApprox.o

        --- Compiled ParOptCompactEigenvalueApprox.cpp successfully ---

mpicxx -fPIC -O3 -I/tmp/tmp_nnp492g/src -c CyParOptProblem.cpp -o CyParOptProblem.o

        --- Compiled CyParOptProblem.cpp successfully ---

/usr/share/miniconda/envs/test/bin/x86_64-conda-linux-gnu-ar rcs /tmp/tmp_nnp492g/lib/libparopt.a ParOptOptions.o ParOptInteriorPoint.o ParOptVec.o ParOptQuasiNewton.o ParOptMMA.o ParOptTrustRegion.o ParOptProblem.o ParOptOptimizer.o ParOptCompactEigenvalueApprox.o CyParOptProblem.o
gmake[1]: Leaving directory '/tmp/tmp_nnp492g/src'
mpicxx -fPIC -shared src/CyParOptProblem.o src/ParOptCompactEigenvalueApprox.o src/ParOptInteriorPoint.o src/ParOptMMA.o src/ParOptOptimizer.o src/ParOptOptions.o src/ParOptProblem.o src/ParOptQuasiNewton.o src/ParOptTrustRegion.o src/ParOptVec.o -L/usr/lib -llapack -lblas -o /tmp/tmp_nnp492g/lib/libparopt.so
 
Installing paropt with pip...
Error: Process completed with exit code 1.
```

--- a/.github/workflows/test_workflow.yml
+++ b/.github/workflows/test_workflow.yml
@@ -226,7 +226,7 @@ jobs:
           echo "============================================================="
           python -m pip_audit
 
-      - name: Slack deprecation warnings
+      - name: Slack audit warnings
         if: steps.audit.outcome == 'failure'
         uses: act10ns/slack@v2.0.0
         with:

--- a/.github/workflows/test_workflow.yml
+++ b/.github/workflows/test_workflow.yml
@@ -7,6 +7,10 @@ on:
   pull_request:
     branches: [ master ]
 
+  # Run the workflow Sundays at 0400 UTC
+  schedule:
+    - cron: '0 4 * * 0'
+
   # Allow running the workflow manually from the Actions tab
   workflow_dispatch:
 

--- a/.github/workflows/test_workflow.yml
+++ b/.github/workflows/test_workflow.yml
@@ -7,6 +7,9 @@ on:
   pull_request:
     branches: [ master ]
 
+  # Allow running the workflow manually from the Actions tab
+  workflow_dispatch:
+
 jobs:
 
   tests:

--- a/.github/workflows/test_workflow.yml
+++ b/.github/workflows/test_workflow.yml
@@ -25,7 +25,7 @@ jobs:
             SCIPY: 1.7
             MPI4PY: true
             PYOPTSPARSE: 'v2.8.3'
-            PAROPT: true
+            # PAROPT: true
             SNOPT: 7.7
 
           # test baseline versions on MacOS
@@ -36,7 +36,7 @@ jobs:
             SCIPY: 1
             MPI4PY: true
             PYOPTSPARSE: 'latest'
-            PAROPT: true
+            # PAROPT: true
             SNOPT: 7.7
 
           # test latest versions
@@ -47,7 +47,7 @@ jobs:
             SCIPY: 1
             MPI4PY: true
             PYOPTSPARSE: 'latest'
-            PAROPT: true
+            # PAROPT: true
             SNOPT: 7.7
 
           # test oldest supported versions
@@ -180,7 +180,7 @@ jobs:
               echo "HSL was requested but source is not available, using default linear solver."
               echo "---------------------------------------------------------------------------"
             fi
-          elif [[ "${{  matrix.LINEAR_SOLVER }}" == "pardiso" ]]; then
+          elif [[ "${{ matrix.LINEAR_SOLVER }}" == "pardiso" ]]; then
             echo "-------------------------------------------------------------------------------"
             echo "Pardiso requires Intel compilers, which are not installed. The build will fail."
             echo "-------------------------------------------------------------------------------"
@@ -209,6 +209,8 @@ jobs:
           testflo --pre_announce --show_skipped .
 
       - name: Audit dependencies
+        id: audit
+        continue-on-error: true
         run: |
           python -m pip install pip-audit
 
@@ -216,6 +218,16 @@ jobs:
           echo "Scan environment for packages with known vulnerabilities"
           echo "============================================================="
           python -m pip_audit
+
+      - name: Slack deprecation warnings
+        if: steps.audit.outcome == 'failure'
+        uses: act10ns/slack@v2.0.0
+        with:
+          webhook-url: ${{ secrets.SLACK_WEBHOOK_URL }}
+          status: 'warning'
+          message: |
+            pip-audit detected vulnerabilities.
+            ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}
 
       - name: Notify slack
         uses: act10ns/slack@v1

--- a/.github/workflows/test_workflow.yml
+++ b/.github/workflows/test_workflow.yml
@@ -32,7 +32,7 @@ jobs:
             SCIPY: 1.7
             MPI4PY: true
             PYOPTSPARSE: 'v2.8.3'
-            # PAROPT: true
+            PAROPT: true
             SNOPT: 7.7
 
           # test baseline versions on MacOS
@@ -43,7 +43,7 @@ jobs:
             SCIPY: 1
             MPI4PY: true
             PYOPTSPARSE: 'latest'
-            # PAROPT: true
+            PAROPT: true
             SNOPT: 7.7
 
           # test latest versions
@@ -54,7 +54,7 @@ jobs:
             SCIPY: 1
             MPI4PY: true
             PYOPTSPARSE: 'latest'
-            # PAROPT: true
+            PAROPT: true
             SNOPT: 7.7
 
           # test oldest supported versions

--- a/build_pyoptsparse.py
+++ b/build_pyoptsparse.py
@@ -89,7 +89,7 @@ build_info = {
         'include_file': 'CoinHslConfig.h'
     },
     'paropt': {
-        'branch': 'v2.0.2',
+        'branch': '',
         'url': 'https://github.com/smdogroup/paropt.git',
         'src_lib_glob': 'libparopt*',
     }
@@ -398,6 +398,8 @@ def run_cmd(cmd_list, do_check=True, raise_error=True)->bool:
     try:
         result = subprocess.run(cmd_list, check=do_check, capture_output=True, text=True)
     except subprocess.CalledProcessError as inst:
+        if opts['verbose'] is True:
+            print(inst.stdout, inst.stderr)
         if raise_error is True:
             raise inst
 
@@ -598,9 +600,12 @@ def git_clone(build_key:str, auto_delete:bool=True):
     note_ok()
     pushd(dir_name)
 
-    # We don't care about the "detached HEAD" warning:
-    run_cmd(cmd_list=['git', 'config', '--local', 'advice.detachedHead', 'false'])
-    run_cmd(cmd_list=['git', 'checkout', '-q', d['branch']])
+    if d["branch"]:
+        # We don't care about the "detached HEAD" warning:
+        run_cmd(cmd_list=['git', 'config', '--local', 'advice.detachedHead', 'false'])
+        note(f'Checking out branch {d["branch"]}')
+        run_cmd(cmd_list=['git', 'checkout', '-q', d['branch']])
+
     return build_dir
 
 def allow_build(build_key:str) -> bool:
@@ -635,7 +640,7 @@ def install_metis_from_src():
     if not allow_build('metis'):
         return
 
-    build_dir = git_clone('metis')
+    os.environ['METIS_DIR'] = git_clone('metis')
     run_cmd(['./get.Metis'])
     os.environ['CFLAGS'] = '-Wno-implicit-function-declaration'
     note("Running configure")
@@ -649,6 +654,7 @@ def install_metis():
     if allow_install_with_conda() and opts['force_build'] is False:
         try:
             install_conda_pkg('metis')
+            os.environ['METIS_DIR'] = os.environ['CONDA_PREFIX']
             return
         except Exception as e:
             try_fallback('METIS', e)
@@ -698,9 +704,15 @@ def install_paropt_from_src():
     Path('Makefile.in.info').rename('Makefile.in')
     make_vars =  [f'PAROPT_DIR={Path.cwd()}']
     if sys_info['sys_name'] == 'Darwin':
-        make_vars.extend(['SO_EXT=dylib', 'SO_LINK_FLAGS=-fPIC -dynamiclib'])
+        make_vars.extend(['SO_EXT=so', 'SO_LINK_FLAGS=-fPIC -dynamiclib -undefined dynamic_lookup',
+                          f'METIS_INCLUDE=-I{os.environ["METIS_DIR"]}/include/',
+                          f'METIS_LIB=-L{os.environ["METIS_DIR"]}/lib/',
+                          '-lmetis'])
     else:
-        make_vars.extend(['SO_EXT=so', 'SO_LINK_FLAGS=-fPIC -shared'])
+        make_vars.extend(['SO_EXT=so', 'SO_LINK_FLAGS=-fPIC -shared',
+                          f'METIS_INCLUDE=-I{os.environ["METIS_DIR"]}/include/',
+                          f'METIS_LIB=-L{os.environ["METIS_DIR"]}/lib/',
+                          '-lmetis'])
 
     make_install(make_args=make_vars, do_install=False)
     pip_install(['./'], pkg_desc='paropt')

--- a/build_pyoptsparse.py
+++ b/build_pyoptsparse.py
@@ -11,6 +11,7 @@ import tempfile
 from colors import *
 from shutil import which
 from packaging.version import Version, parse
+import setuptools
 
 # Default options that the user can change with command line switches
 opts = {
@@ -251,7 +252,6 @@ def process_command_line():
     opts['verbose'] = args.verbose
     opts['uninstall'] = args.uninstall
 
-
 def announce(msg:str):
     """
     Print an important message in color with a line above and below.
@@ -471,7 +471,6 @@ def run_conda_cmd(cmd_args):
     cmd_list = [opts['conda_cmd']]
     cmd_list.extend(cmd_args)
     return run_cmd(cmd_list)
-
 
 def pip_install(pip_install_args, pkg_desc='packages'):
     """
@@ -915,15 +914,21 @@ def install_pyoptsparse_from_src():
     if opts['build_pyoptsparse'] is True:
         patch_pyoptsparse_src()
 
-        # `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
-        # of the deprecation of `distutils` itself. It will be removed for
-        # Python >= 3.12. For older Python versions it will remain present.
-        # It is recommended to use `setuptools < 60.0` for those Python versions.
-        # For more details, see:
-        # https://numpy.org/devdocs/reference/distutils_status_migration.html
-        pip_install(pip_install_args=['setuptools<66.0'], pkg_desc='setuptools')
-
-        pip_install(pip_install_args=['--no-cache-dir', './'], pkg_desc='pyoptsparse')
+        python_ver = Version(platform.python_version())
+        stools_ver = Version(setuptools.__version__)
+        if python_ver < Version('3.12.0') and stools_ver >= Version('66.0'):
+            # `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
+            # of the deprecation of `distutils` itself. It will be removed for
+            # Python >= 3.12. For older Python versions it will remain present.
+            # It is recommended to use `setuptools < 60.0` for those Python versions.
+            # For more details, see:
+            # https://numpy.org/devdocs/reference/distutils_status_migration.html
+            pip_install(pip_install_args=['setuptools<66.0'], pkg_desc='setuptools<66.0')
+            pip_install(pip_install_args=['--no-cache-dir', './'], pkg_desc='pyoptsparse')
+            pip_install(pip_install_args=[f'setuptools=={stools_ver}'],
+                        pkg_desc='previous version of setuptools')
+        else:
+            pip_install(pip_install_args=['--no-cache-dir', './'], pkg_desc='pyoptsparse')
     else:
         announce('Not building pyOptSparse by request')
         if opts['include_ipopt'] is True:

--- a/build_pyoptsparse.py
+++ b/build_pyoptsparse.py
@@ -89,7 +89,7 @@ build_info = {
         'include_file': 'CoinHslConfig.h'
     },
     'paropt': {
-        'branch': '',
+        'branch': 'v2.1.4',
         'url': 'https://github.com/smdogroup/paropt.git',
         'src_lib_glob': 'libparopt*',
     }

--- a/build_pyoptsparse.py
+++ b/build_pyoptsparse.py
@@ -704,7 +704,7 @@ def install_paropt_from_src():
     Path('Makefile.in.info').rename('Makefile.in')
     make_vars =  [f'PAROPT_DIR={Path.cwd()}']
     if sys_info['sys_name'] == 'Darwin':
-        make_vars.extend(['SO_EXT=so', 'SO_LINK_FLAGS=-fPIC -dynamiclib -undefined dynamic_lookup',
+        make_vars.extend(['SO_EXT=dylib', 'SO_LINK_FLAGS=-fPIC -dynamiclib -undefined dynamic_lookup',
                           f'METIS_INCLUDE=-I{os.environ["METIS_DIR"]}/include/',
                           f'METIS_LIB=-L{os.environ["METIS_DIR"]}/lib/',
                           '-lmetis'])

